src/biz/workspace/publish.rs

```diff
@@ -0,0 +1,156 @@
+use app_error::AppError;
+use database_entity::dto::{PublishCollabItem, PublishInfo};
+use sqlx::PgPool;
+use uuid::Uuid;
+
+use database::{
+  publish::{
+    delete_published_collabs, insert_or_replace_publish_collabs, select_publish_collab_meta,
+    select_published_collab_blob, select_published_collab_info,
+    select_user_is_collab_publisher_for_all_views, select_workspace_publish_namespace,
+    select_workspace_publish_namespace_exists, update_workspace_publish_namespace,
+  },
+  workspace::select_user_is_workspace_owner,
+};
+
+use super::ops::check_workspace_owner;
+
+pub async fn publish_collabs(
+  pg_pool: &PgPool,
+  workspace_id: &Uuid,
+  publisher_uuid: &Uuid,
+  publish_items: Vec<PublishCollabItem<serde_json::Value, Vec<u8>>>,
+) -> Result<(), AppError> {
+  for publish_item in &publish_items {
+    check_collab_publish_name(publish_item.meta.publish_name.as_str())?;
+  }
+  insert_or_replace_publish_collabs(pg_pool, workspace_id, publisher_uuid, publish_items).await?;
```

Comment:
Since there's no explicit Postgres transaction intiialization I can only assume that every sqlx query executed here will be wrapped and committed independently.

Another issue is that publish_items has no upper bound check. We basically copy unlimited number of collabs of unlimited size. With no timeouts.

Reply:
there's only 1 sqlx operation for this api call, which is the `insert_or_replace_publish_collabs`. so there's no need for transaction. as mentioned above, upper bound is checked when api is call before it reaches the this part